src/vos/vos_io.c

```diff
@@ -2080,47 +1959,33 @@ vos_update_end(daos_handle_t ioh, uint32_t pm_ver, daos_key_t *dkey, int err,
 	/** Now that we are past the existence checks, ensure there isn't a
 	 * read conflict
 	 */
-	if (ioc->ic_read_conflict) {
+	if (vos_ts_set_check_conflict(ioc->ic_ts_set, ioc->ic_epr.epr_hi)) {
 		err = -DER_TX_RESTART;
 		goto abort;
 	}
 
-	if (dth != NULL) {
-		struct dtx_rsrvd_uint	*dru;
-
-		dru = &dth->dth_rsrvds[dth->dth_rsrvd_cnt++];
-		dru->dru_scm = ioc->ic_rsrvd_scm;
-		ioc->ic_rsrvd_scm = NULL;
-
-		D_INIT_LIST_HEAD(&dru->dru_nvme);
-		d_list_splice_init(&ioc->ic_blk_exts, &dru->dru_nvme);
-	} else {
-		/* Publish SCM reservations */
-		err = vos_publish_scm(ioc->ic_cont, ioc->ic_rsrvd_scm, true);
-		if (err == 0)
-			/* Publish NVMe reservations */
-			err = vos_publish_blocks(ioc->ic_cont,
-						 &ioc->ic_blk_exts, true,
-						 VOS_IOS_GENERIC);
-	}
-
 abort:
-	err = vos_tx_end(dth, umem, err);
+	err = vos_tx_end(ioc->ic_cont, dth, &ioc->ic_rsrvd_scm,
+			 &ioc->ic_blk_exts, err);
 
 out:
-	if (err != 0) {
-		update_cancel(ioc);
-	} else {
+	if (err == 0) {
 		if (daes != NULL)
 			vos_dtx_post_handle(ioc->ic_cont, daes,
 					    dth->dth_dti_cos_count, false);
 		vos_dedup_process(vos_cont2pool(ioc->ic_cont),
 				  &ioc->ic_dedup_entries, false);
+	} else {
+		update_cancel(ioc);
```

Comment:
The SCM & blocks publish code for non-DTX case has been move into vos_tx_end(), should we move the cancel part into vos_tx_end() as well?

Reply:
I debated here because the update_cancel is now only used if we are in vmem mode

Reply:
Right, I did that